File: pdoo/java/Deepspace/Dice.java
/**
 * @author Manuel Gachs Ballegeer
 */
package Deepspace;

import java.util.Random;

/**
 * @brief Dado del juego
 */
class Dice {
    private final float NHANGARSPROB;
    private final float NSHIELDSPROB;
    private final float NWEAPONSPROB;
    private final float FIRSTSHOTPROB;
    private final float EXTRAEFFICIENCYPROB;
    private final Random generator;
    
    Dice() {
        NHANGARSPROB = 0.25f;
        NSHIELDSPROB = 0.25f;
        NWEAPONSPROB = 0.33f;
        FIRSTSHOTPROB = 0.5f;
        EXTRAEFFICIENCYPROB = 0.8f;
        generator = new Random();
    }
    public int initWithNHangars() {
        float prob = generator.nextFloat();
        
        if (prob<NHANGARSPROB)
            return 0;
        else
            return 1;
    }
    public int initWithNWeapons() {
        float prob = generator.nextFloat();
        
        if (prob<NWEAPONSPROB)
            return 1;
        else if (NWEAPONSPROB<=prob && prob<2*NWEAPONSPROB)
            return 3;
        else
            return 2;
    }
    public int initWithNShields() {
        float prob = generator.nextFloat();
        
        if (prob<NSHIELDSPROB)
            return 0;
        else
            return 1;
    }
    public int whoStarts(int nPlayers) {
        return generator.nextInt(nPlayers-1);
    }
    public GameCharacter firstShot() {
        float prob = generator.nextFloat();
        
        if (prob<FIRSTSHOTPROB)
            return GameCharacter.SPACESTATION;
        else
            return GameCharacter.ENEMYSTARSHIP;
    }
    public boolean spaceStationMoves(float speed) {
        float prob = generator.nextFloat();
        
        return prob<speed;
    }
    
    public boolean extraEfficiency() {
        float prob = generator.nextFloat();
        
        if (prob<EXTRAEFFICIENCYPROB)
            return true;
        else
            return false;
    }
}

File: pdoo/ruby/dice.rb
#encoding:utf-8

require_relative "./gamecharacter.rb"

module Deepspace
  # Clase dado
  class Dice
    def initialize
      @NHANGARSPROB = 0.25
      @NSHIELDSPROB = 0.25 
      @NWEAPONSPROB = 0.33
      @FIRSTSHOTPROB = 0.25
      @EXTRAEFFICIENCYPROB = 0.8
      @GENERATOR = Random.new()
    end

    def initWithNHangars
      prob = @GENERATOR.rand(1.0)
      if prob<@NHANGARSPROB
        return 0
      else
        return 1
      end
    end

    def initWithNWeapons
      prob = @GENERATOR.rand(1.0)
      if prob<@NWEAPONSPROB
        return 1
      elsif @NWEAPONSPROB<=prob and prob<2*@NWEAPONSPROB
        return 2
      else
        return 3
      end
    end
    
    def initWithNShields
      prob = @GENERATOR.rand(1.0)
      if prob<@NSHIELDSPROB
        return 0
      else
        return 1
      end
    end

    def whoStarts(n)
      @GENERATOR.rand(n-1)
    end

    def firstShot
      prob = @GENERATOR.rand(1.0)
      if prob<@FIRSTSHOTPROB
        return GameCharacter::SPACESTATION
      else
        return GameCharacter::ENEMYSTARSHIP
      end
    end

    def spaceStationMoves(speed)
      prob = @GENERATOR.rand(1.0)
      if prob<speed
        return true
      else
        return false
      end
    end

    def extraEfficiency
      prob = @GENERATOR.rand(1.0)
      if prob<@EXTRAEFFICIENCYPROB
        return true
      else
        return false
      end
    end
  end
end
